### PR TITLE
[CH 7317] One Way Synonyms Support

### PR DESCRIPTION
**Pull request by jjl014**
* Update `makeUrl` to take an options object
* Add support for One Way Synonyms
* Add tests around One Way Synonyms

Notes:
Still missing a test case around having an empty `child_phrases` array, but should be ready overall and mergeable. Can always add it in later when the issue is resolved.

Discussion about the above here: https://constructor.slack.com/archives/CA44G41HA/p1568830253012600

### CODE REVIEW

File: lib/constructorio.js [hunk 1]
```diff
@@ -384,6 +385,121 @@ ConstructorIO.prototype = {
...
+      method: 'DELETE',
+      url: this.makeUrl('one_way_synonyms', { basePath: 'v2' }),
+      auth: this.makeAuthToken(),
+    }, (error, response) => {
+      handleServerResponse(error, response, callback);
+    });
+  },
+
+  /**
+   * @description Remove a one way synonym for supplied parent phrase.
```
Comment (sblaurock) [inline lib/constructorio.js]: Super small thing but none of the other comments end in a period (.)

File: test/constructorio-synonyms-one-way.js [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+    it('should return an error when not supplying a child phrases property', (done) => {
+      const constructorio = new Constructorio(testConfig);
+
+      constructorio.addOneWaySynonym({}, (err, response) => {
+        expect(err).to.be.an('object');
+        expect(err).to.have.property('message', 'child_phrases is a required field of type list');
+        expect(response).to.be.undefined;
+        done();
+      });
+    });
```
Comment (sblaurock) [inline test/constructorio-synonyms-one-way.js]: Maybe a test for when a `phrase` isn't supplied?

File: test/constructorio-synonyms-one-way.js [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+  });
+
+  describe('modifyOneWaySynonym', () => {
+    const oneWaySynonymPhrase = 'bread';
+
+    before((done) => {
+      // Create test one way synonyms for use in tests
+      setTimeout(() => {
+        addTestOneWaySynonym(oneWaySynonymPhrase).then(done).catch(done);
+      }, 100);
```
Comment (sblaurock) [inline test/constructorio-synonyms-one-way.js]: Do we need the setTimeout here?

File: test/constructorio-synonyms-one-way.js [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+        expect(response).to.be.an('object');
+        expect(response).to.have.property('one_way_synonym_relations').an('array').length(0);
+        done();
+      });
+    });
+
+    it('should retrieve a listing of one way synonyms when supplying num results per page property', (done) => {
+      const constructorio = new Constructorio(testConfig);
+
+      // Note: Result set is not being checked for match as groups can take many seconds to be indexed / returned
```
Comment (sblaurock) [inline test/constructorio-synonyms-one-way.js]: Is this comment still true? If so, we should change "groups" to "synonyms".

File: test/constructorio-synonyms-one-way.js [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+        phrase: 1,
+      }, (err, response) => {
+        expect(err).to.be.undefined;
+        expect(response).to.be.an('object');
+        expect(response).to.have.property('one_way_synonym_relations').an('array').length(0);
+        done();
+      });
+    });
+
+    it('should return no results when not supplying a phrase', (done) => {
```
Comment (sblaurock) [inline test/constructorio-synonyms-one-way.js]: Surprised this doesn't throw an error!

File: test/constructorio-synonyms-one-way.js [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+      Promise.all(removePromiseList).then(() => done()).catch(() => done());
+    });
+
+    it('should start removal of all one way synonyms', (done) => {
+      const constructorio = new Constructorio(testConfig);
+
+      constructorio.removeOneWaySynonyms({}, (err, response) => {
+        expect(err).to.be.undefined;
+        expect(response).to.be.an('object');
+        expect(response).to.have.property('message', 'We\'ve started deleting all of your synonym groups. This may take some time to complete.');
```
Comment (sblaurock) [inline test/constructorio-synonyms-one-way.js]: "synonym groups"? language issue on back-end, or does it actually say this?

File: lib/constructorio.js [hunk 7]
```diff
@@ -384,6 +385,121 @@ ConstructorIO.prototype = {
...
+    });
+  },
+
+  /**
+   * @description Retrieves a one way synonym for supplied parent phrase.
+   */
+  getOneWaySynonym(params, callback) {
+    const json = clonedeep(params);
+    const { phrase } = json;
+    delete json.group_id;
```
Comment (sblaurock) [inline lib/constructorio.js]: should be `phrase`

File: test/constructorio-synonyms-one-way.js [hunk 8]
```diff
@@ -0,0 +1,587 @@
...
+        expect(response).to.be.an('object');
+        expect(response).to.have.property('one_way_synonym_relations').to.be.an('array').length(3);
+        done();
+      });
+    });
+
+    it('should retrieve a listing of one way synonyms when supplying a phrase property', (done) => {
+      const constructorio = new Constructorio(testConfig);
+
+      // Note: Result set is not being checked for match as groups can take many seconds to be indexed / returned
```
Comment (sblaurock) [inline test/constructorio-synonyms-one-way.js]: Is this still true? If so, should be "synonyms" rather than "groups".

File: lib/constructorio.js (hunk 1 above)
Comment (jjl014) [reply lib/constructorio.js]: Updated!

File: test/constructorio-synonyms-one-way.js (hunk 2 above)
Comment (jjl014) [reply test/constructorio-synonyms-one-way.js]: Added!

File: test/constructorio-synonyms-one-way.js (hunk 3 above)
Comment (jjl014) [reply test/constructorio-synonyms-one-way.js]: Nope, good catch!

File: test/constructorio-synonyms-one-way.js (hunk 4 above)
Comment (jjl014) [reply test/constructorio-synonyms-one-way.js]: Yup, updated!

File: test/constructorio-synonyms-one-way.js (hunk 5 above)
Comment (jjl014) [reply test/constructorio-synonyms-one-way.js]: I added a check on the client side since it ends up treating `"undefined"` as the phrase if it isn't provided, which is not right... So it will now throw an error if a phrase isn't provided.

File: test/constructorio-synonyms-one-way.js (hunk 6 above)
Comment (jjl014) [reply test/constructorio-synonyms-one-way.js]: Yea, it actually does say that. And it seems like they don't differentiate between groups and one way synonyms. https://github.com/Constructor-io/autocomplete/blob/6b6441a0a7e4ef76e7cad0116f833fb682586655/common/src/common/api/messages.py#L23

File: lib/constructorio.js (hunk 7 above)
Comment (jjl014) [reply lib/constructorio.js]: Good catch, updated!

File: test/constructorio-synonyms-one-way.js (hunk 8 above)
Comment (jjl014) [reply test/constructorio-synonyms-one-way.js]: Updated!

**Comment by sblaurock** [comment]
Nice work. Ship it!
